File: src/backend/schemas/community.py
"""Community Safety Hub API schemas (Slice F).

Request/response contracts for the public Community Safety Hub and the
SYSTEM_ADMIN CMS management endpoints. Title/body are plain text only — see
the XSS/rendering note in ``services/community_content`` and
``system-wiki/security/security-baseline.md``. The frontend MUST render these
fields as React text (auto-escaped) and MUST NOT use ``dangerouslySetInnerHTML``.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field

CommunityContentType = Literal["SAFETY_ARTICLE", "ANNOUNCEMENT", "EVENT"]
CommunityLanguage = Literal["en", "uk"]


class CommunityContentCreate(BaseModel):
    """Request body for POST /api/admin/community (create draft)."""

    content_type: CommunityContentType
    title_en: str = Field(..., min_length=1)
    body_en: str = Field(..., min_length=1)
    title_uk: str | None = None
    body_uk: str | None = None
    metadata_json: dict[str, Any] | None = None
    slug: str | None = Field(default=None, min_length=1, max_length=255)
    expires_at: datetime | None = None
    urgent_banner: bool = False
    last_reviewed_at: datetime | None = None


class CommunityContentUpdate(BaseModel):
    """Request body for PATCH /api/admin/community/{content_id} (edit DRAFT)."""

    title_en: str | None = Field(default=None, min_length=1)
    body_en: str | None = Field(default=None, min_length=1)
    title_uk: str | None = Field(default=None, min_length=1)
    body_uk: str | None = Field(default=None, min_length=1)
    metadata_json: dict[str, Any] | None = None
    slug: str | None = Field(default=None, min_length=1, max_length=255)
    expires_at: datetime | None = None
    urgent_banner: bool | None = None
    last_reviewed_at: datetime | None = None


class CommunityContentPublish(BaseModel):
    """Request body for POST /api/admin/community/{content_id}/publish."""

    title_en: str = Field(..., min_length=1)
    body_en: str = Field(..., min_length=1)
    title_uk: str | None = None
    body_uk: str | None = None
    metadata_json: dict[str, Any] | None = None
    expires_at: datetime | None = None
    urgent_banner: bool = False
    last_reviewed_at: datetime | None = None


class CommunityContentItem(BaseModel):
    """Public projection of one published content item (language-resolved)."""

    content_id: str
    slug: str
    content_type: CommunityContentType
    title: str
    body: str
    language: CommunityLanguage
    urgent_banner: bool
    expires_at: datetime | None = None
    metadata_json: dict[str, Any] | None = None
    last_reviewed_at: datetime | None = None
    updated_at: datetime | None = None


class CommunityHubResponse(BaseModel):
    """Response body for GET /api/community/hub."""

    items: list[CommunityContentItem]
    urgent_banner: CommunityContentItem | None = None


class CommunityContentDetailResponse(BaseModel):
    """Response body for GET /api/community/{slug}."""

    item: CommunityContentItem


class CommunityContentAdminItem(BaseModel):
    """Full latest-version projection for the SYSTEM_ADMIN CMS editor."""

    content_id: str
    slug: str
    content_type: CommunityContentType
    lifecycle_status: str
    title_en: str
    title_uk: str | None = None
    body_en: str
    body_uk: str | None = None
    metadata_json: dict[str, Any] | None = None
    expires_at: datetime | None = None
    urgent_banner: bool
    last_reviewed_at: datetime | None = None
    row_version: int


class CommunityContentCreateResponse(BaseModel):
    """Response body for POST /api/admin/community."""

    content_id: str
    lifecycle_status: str = "DRAFT"


class CommunityContentActionResponse(BaseModel):
    """Generic response for publish/archive/update actions."""

    content_id: str
    lifecycle_status: str

File: src/frontend/src/lib/api/community.ts
import { ApiParseError } from '@/lib/validation';
import { publicApiFetch } from './public-transport';

export type CommunityLanguage = 'en' | 'uk';
export type CommunityContentType = 'SAFETY_ARTICLE' | 'ANNOUNCEMENT' | 'EVENT';

export interface CommunityContentItem {
  content_id: string;
  slug: string;
  content_type: CommunityContentType;
  title: string;
  body: string;
  language: CommunityLanguage;
  urgent_banner: boolean;
  expires_at: string | null;
  metadata_json: Record<string, unknown> | null;
  last_reviewed_at: string | null;
  updated_at: string | null;
}

export interface CommunityHubResponse {
  items: CommunityContentItem[];
  urgent_banner: CommunityContentItem | null;
}

export interface CommunityContentDetailResponse {
  item: CommunityContentItem;
}

export function fetchCommunityHub(options: {
  language?: CommunityLanguage;
  type?: CommunityContentType;
} = {}): Promise<CommunityHubResponse> {
  const params = new URLSearchParams({ language: options.language ?? 'en' });
  if (options.type) params.set('type', options.type);
  return publicApiFetch<unknown>(`/community/hub?${params.toString()}`).then(parseHubResponse);
}

export function fetchCommunityContent(
  slug: string,
  language: CommunityLanguage = 'en',
): Promise<CommunityContentDetailResponse> {
  return publicApiFetch<unknown>(
    `/community/${encodeURIComponent(slug)}?language=${language}`,
  ).then(parseDetailResponse);
}

function parseHubResponse(value: unknown): CommunityHubResponse {
  if (!isRecord(value) || !Array.isArray(value.items)) return invalidResponse();
  const items = value.items.map(parseItem);
  const urgentBanner = value.urgent_banner === null ? null : parseItem(value.urgent_banner);
  return { items, urgent_banner: urgentBanner };
}

function parseDetailResponse(value: unknown): CommunityContentDetailResponse {
  if (!isRecord(value) || !('item' in value)) return invalidResponse();
  return { item: parseItem(value.item) };
}

function parseItem(value: unknown): CommunityContentItem {
  if (!isRecord(value)) return invalidResponse();
  const contentType = value.content_type;
  const language = value.language;
  if (
    typeof value.content_id !== 'string' || typeof value.slug !== 'string'
    || (contentType !== 'SAFETY_ARTICLE' && contentType !== 'ANNOUNCEMENT' && contentType !== 'EVENT')
    || (language !== 'en' && language !== 'uk')
    || typeof value.title !== 'string' || typeof value.body !== 'string'
    || typeof value.urgent_banner !== 'boolean'
    || !isNullableTimestamp(value.expires_at)
    || !isNullableMetadata(value.metadata_json)
    || !isNullableTimestamp(value.last_reviewed_at)
    || !isNullableTimestamp(value.updated_at)
  ) return invalidResponse();
  return {
    content_id: value.content_id,
    slug: value.slug,
    content_type: contentType,
    title: value.title,
    body: value.body,
    language,
    urgent_banner: value.urgent_banner,
    expires_at: value.expires_at,
    metadata_json: value.metadata_json,
    last_reviewed_at: value.last_reviewed_at,
    updated_at: value.updated_at,
  };
}

function isNullableTimestamp(value: unknown): value is string | null {
  return value === null || (typeof value === 'string' && Number.isFinite(Date.parse(value)));
}

function isNullableMetadata(value: unknown): value is Record<string, unknown> | null {
  return value === null || isRecord(value);
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function invalidResponse(): never {
  throw new ApiParseError('Invalid community API response shape', 200);
}
